File: python/gui.py
import numpy as np
import cv2

# Cityscapes class colors
# Ref: https://github.com/mcordts/cityscapesScripts/blob/master/cityscapesscripts/helpers/labels.py
COLORS = [
    (0, 0, 0),
    (0, 0, 0),
    (0, 0, 0),
    (0, 0, 0),
    (0, 0, 0),
    (111, 74, 0),
    (81, 0, 81),
    (128, 64, 128),
    (244, 35, 232),
    (250, 170, 160),
    (230, 150, 140),
    (70, 70, 70),
    (102, 102, 156),
    (190, 153, 153),
    (180, 165, 180),
    (150, 100, 100),
    (150, 120, 90),
    (153, 153, 153),
    (153, 153, 153),
    (250, 170, 30),
    (220, 220, 0),
    (107, 142, 35),
    (152, 251, 152),
    (70, 130, 180),
    (220, 20, 60),
    (255, 0, 0),
    (0, 0, 142),
    (0, 0, 70),
    (0, 60, 100),
    (0, 0, 90),
    (0, 0, 110),
    (0, 80, 100),
    (0, 0, 230),
    (119, 11, 32),
    (0, 0, 142),
]


def get_max_text_size(textlines, fontface, scale, thickness):
    """Extracts the size of the widest text line"""
    max_size, _ = cv2.getTextSize(textlines[0], fontface, scale, thickness)
    for s in textlines[1:]:
        size, _ = cv2.getTextSize(s, fontface, scale, thickness)
        if size[0] > max_size[0]:
            max_size = size
    return max_size


def add_text(image, x, y, textlines):
    """Draws a popup with metadata contained in textlines"""

    # Style parameters
    fontface = cv2.FONT_HERSHEY_SIMPLEX
    scale = 0.4
    thickness = 1
    line_padding = 30
    padding_x = 5
    padding_y = 30

    n = len(textlines)
    h, w = image.shape[:2]

    text_width, text_height = get_max_text_size(textlines, fontface, scale, thickness)

    # Rectangle height and width
    rect_height = n * text_height + (n - 1) * line_padding
    rect_width = text_width + 2 * padding_x
    # Top-left coordinate of rect (ensure we don't draw out of frame)
    y_top = min(h - rect_height, max(0, y - padding_y))
    x_left = min(w - rect_width, max(0, x - padding_x))
    # Upper left and bottom right corners of the rectangle
    upper_left = (x_left, y_top)
    bottom_right = (x_left + rect_width, y_top + rect_height)
    cv2.rectangle(image, upper_left, bottom_right, (0, 0, 0), cv2.FILLED)
    # Set text coordinate based on upper left corner of rect
    x = x_left + padding_x
    y = y_top + padding_y
    for line in textlines:
        cv2.putText(
            image, line, (x, y), fontFace=fontface, fontScale=scale, color=(255, 255, 255), thickness=thickness
        )
        y += line_padding


def draw_contours(image, segments):
    """Draws segment contours"""
    for p in segments:
        pixels = np.array(list(p.contour_pixels))
        image[pixels[:, 0], pixels[:, 1]] = COLORS[p.label % len(COLORS)]


def visualize(image, segments, labels):
    """Main GUI method. Opens a frame and waits for commands"""

    def mouse_callback(event, x, y, flags, param):
        """Callback method for mouse clicks. Triggers a popup in case of left clicks."""
        if event == cv2.EVENT_LBUTTONUP:
            # Find segment of which the point is a member
            segment = [s for s in segments if s.seen((y, x))][0]
            add_text(
                image,
                x,
                y,
                [
                    f"Point: ({x}, {y})",
                    f"Name: {labels[segment.label]}",
                    f"Label: {segment.label}",
                    f"Score: {segment.pixels[(y, x)].score}",
                    f"Segment Score: {segment.score}",
                    f"Segment ID: {segment.id}",
                ],
            )

    draw_contours(image, segments)

    cv2.namedWindow("image")
    cv2.setMouseCallback("image", mouse_callback)

    # Save image clone to enable resets
    image_clone = image.copy()

    # Loop until 'q' or 'c' is pressed
    while True:

        cv2.imshow("image", image)

        # Wait for keypress
        key = cv2.waitKey(1) & 0xFF

        # Press 'r' to reset the frame
        if key == ord("r"):
            image = image_clone.copy()

        # Press 'q' or 'c' to break from the loop
        elif key in [ord("q"), ord("c")]:
            break

File: c++/src/gui.cpp
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include "gui.hpp"
#include <iostream>
#include <iomanip>
#include <string>

using namespace std;
using namespace cv;


/**
    Helper method which converts input to string format

    @param t : input
    @return string representation of t
*/
template <typename T>
string toStr(const T& t) {
   ostringstream os;
   os << setprecision(2) << t;
   return os.str();
}


/**
    Returns size of the widest text line in textlines
    @return size of widest text lines
*/
Size getMaxTextSize(vector<string> lines, int fontface, double scale, int thickness, int baseline) {
  Size maxSz;
  maxSz.width = -1;
  for(string s : lines) {
    Size sz = getTextSize(s, fontface, scale, thickness, &baseline);
    if(sz.width > maxSz.width)
      maxSz = sz;
  }
  return maxSz;
}


/**
    Draws a black rectangle with metadata (specified by 'lines') at coordinate (x, y)

    @param image
    @param x : x coordinate of point
    @param y : y coordinate of point
    @param lines : vector of text lines (metadata)
*/
void addTextLines(Mat image, int x, int y, vector<string> lines) {

  // Style parameters
  int fontface = FONT_HERSHEY_SIMPLEX;
  double scale = 0.4;
  int thickness = 1;
  int baseline = 0;
  int linePadding = 30;
  int paddingX = 5;
  int paddingY = 30;

  int imgHeight = image.rows;
  int imgWidth = image.cols;
  int n = lines.size();

  // Find the widest text line
  Size text = getMaxTextSize(lines, fontface, scale, thickness, baseline);
  // Height and width of the background rectangle
  int rectHeight = n * text.height + (n - 1) * linePadding;
  int rectWidth = text.width + 2 * paddingX;
  // Ensure we don't draw rect out of frame
  int yTop = min(imgHeight - rectHeight, max(0, y - paddingY));
  int xLeft = min(imgWidth - rectWidth, max(0, x - paddingX));
  // Upper left and bottom right corners of the rectangle
  Point upperLeft = Point(xLeft, yTop);
  Point bottomRight = Point(
    xLeft + rectWidth,
    yTop + rectHeight
  );
  // Draw background rectangle
  rectangle(image, upperLeft, bottomRight, Scalar(0, 0, 0), CV_FILLED);

  // Set text coordinates relative to upper left coordinate of rectangle
  x = xLeft + paddingX;
  y = yTop + paddingY;
  // Draw text lines on top of the rectangle
  for(int i = 0; i < lines.size(); i++) {
    putText(
        image, lines[i], Point(x, y), fontface, scale, Scalar(255, 255, 255), thickness, 8
    );
    y += linePadding;
  }
  // Display image
  imshow("image", image);
}


/**
    Callback function that is triggered for left mouse clicks. Extracts relevant metadata
    for the coordinate of the click and calls addTextLines to display the data.

    @param event : event type
    @param x : x coordinate of click
    @param y : y coordinate of click
    @param flags : ignored
    @param userData : CallbackParams object containing image, segments and label names
*/
void mouseCallback(int event, int x, int y, int flags, void* userdata)
{
     if  ( event == EVENT_LBUTTONUP )
     {
        // Extract image and segments
        CallbackParams* params = (CallbackParams*) userdata;
        Mat image = params->image;
        vector<Segment*> segments = params->segments;
        vector<string> labels = params->labels;

        // Find the segment that was clicked
        Segment* s = segmentCorrespondingToPoint(segments, x, y);

        cout  << "Left button of the mouse is clicked - position (" << x
              << ", " << y << ") - label " << s->getLabel()
              << " - pid " << s->getId() << endl;

        // Define metadata to be visualized
        vector<string> lines;
        lines.push_back(toStr("Point (") + toStr(x) + toStr(", ") + toStr(y) + toStr(")"));
        lines.push_back(toStr("Name ") + toStr(labels[s->getLabel()]));
        lines.push_back(toStr("Label ") + toStr(s->getLabel()));
        lines.push_back(toStr("Score ") + toStr(s->getPixel(x, y).score));
        lines.push_back(toStr("Segment Score ") + toStr(s->getScore()));
        lines.push_back(toStr("Segment ID ") + toStr(s->getId()));

        // Draw metadata
        addTextLines(image, x, y, lines);
     }
}


/**
    Draws each segment contour to the image

    @param image
    @param segments : image segments
*/
void drawContours(Mat& image, vector<Segment*> segments) {
  for(Segment* s : segments)
    for(Pixel p : s->getCountour())
      for(int i = 0; i < 3; i++)
        image.at<Vec3b>(Point(p.x, p.y))[i] = COLORS[s->getLabel() % COLORS.size()][i];
}


/**
    Main GUI method. Calls other methods to draw image (with contours)
    and sets up callback for mouse click events. Then waits for key press to reset the frame
    or to the close the GUI.

    @param image
    @param segments : image segments
    @param labels : class names
*/
void visualize(Mat image, vector<Segment*> segments, vector<string> labels) {

  // Draw contours of segments
  drawContours(image, segments);

  // Define parameters for mouse click events
  CallbackParams params;
  params.image = image;
  params.segments = segments;
  params.labels = labels;

  // Create a clone to enable frame resets
  Mat imageCopy = image.clone();

  namedWindow("image", 1);
  setMouseCallback("image", mouseCallback, &params);

  while(true){

     imshow("image", params.image);

     // Wait for key command
     char key = waitKey(1) & 0xFF;

     // Press 'r' to reset the frame
     if(key == 'r')
         params.image = imageCopy.clone();
     // Press 'c' or 'q' to break from the loop
     else if(key == 'c' || key == 'q')
         break;
  }
}
